language-support/ts/daml-ledger/index.ts

```diff
@@ -487,28 +495,39 @@ class Ledger {
     template: Template<T, K, I>,
     endpoint: string,
     request: unknown,
+    reconnectRequest: () => unknown,
     init: State,
     change: (state: State, events: readonly Event<T, K, I>[]) => State,
   ): Stream<T, K, I, State> {
     const protocols = ['jwt.token.' + this.token, 'daml.ws.auth'];
-    const ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
-    let isLive = false;
+    let ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
+    let isLiveSince: undefined | number = undefined;
+    let lastOffset: undefined | string = undefined;
     let state = init;
+    let isReconnecting: boolean = false;
     const emitter = new EventEmitter();
-    ws.addEventListener('open', () => {
-      ws.send(JSON.stringify(request));
-    });
-    ws.addEventListener('message', event => {
+    const onOpen = (): void => {
+      if (isReconnecting) {
+          ws.send(JSON.stringify({ 'offset': lastOffset }));
+          ws.send(JSON.stringify(reconnectRequest()));
+      } else
+        ws.send(JSON.stringify(request));
```

Comment:
I'm a big fan of always adding curly braces to simply rule out a big class of potential mistakes, particularly when people start adding code.
```suggestion
      } else {
        ws.send(JSON.stringify(request));
      }
```